### PR TITLE
metrics: Sample metrics periodically and dump samples as part of sysdump

### PR DESCRIPTION
**Pull request by joamaki**
This refactors the disabling of metrics to happen at the `Collect()` time and keeps all registered collectors to the side.
All registered collectors are then sampled periodically which gives us a glimpse to the past two hours:

```
root@kind-worker:/home/cilium# cilium-dbg shell -- metrics -s statedb_table_rev
Metric                          Labels                     5min     30min   60min   120min
cilium_statedb_table_revision   table=backends             31       0       0       0
cilium_statedb_table_revision   table=ciliumenvoyconfigs   0        0       0       0
cilium_statedb_table_revision   table=devices              14       0       0       0
cilium_statedb_table_revision   table=dynamic-features     0        0       0       0
cilium_statedb_table_revision   table=frontends            90       0       0       0
cilium_statedb_table_revision   table=health               1.028k   0       0       0
cilium_statedb_table_revision   table=ipsets               0        0       0       0
cilium_statedb_table_revision   table=l2-announce          0        0       0       0
cilium_statedb_table_revision   table=mtu                  2        0       0       0
cilium_statedb_table_revision   table=node-addresses       14       0       0       0
cilium_statedb_table_revision   table=routes               49       0       0       0
cilium_statedb_table_revision   table=services             12       0       0       0
cilium_statedb_table_revision   table=sysctl               68       0       0       0
```

The sampling is super conservative. The interval is set to 5 minutes, the collectors are invoked sequentially rather than in parallel as is usual with prometheus, and the number of samples per metric is capped to 24 (giving us 2 hour timespan). Also the number of metrics (e.g. "name" + "labels") that we sample is capped to 2000, which limits the maximum memory usage to around 1MB (`TestSamplerMaxMemoryUsage`).

Since the metrics script command now implements same and more as the metrics API endpoint and "cilium-dbg metrics list", this PR removes the metrics API endpoint and reimplements "cilium-dbg metrics list" as a wrapper for the shell command (more about the cilium shell: https://github.com/cilium/cilium/pull/35154). All the same functionality is retained (though now it shows all metrics and not just enabled ones):

```
root@kind-worker:/home/cilium# cilium-dbg metrics list --match-pattern cilium_version
Metric           Labels                                                            Value
cilium_version   arch="linux/amd64" revision="3c6d822f9cfc" version="1.17.0-dev"   0.000000

root@kind-control-plane:/home/cilium# cilium-dbg metrics list -o json -p go_threads
[
  {
    "name": "go_threads",
    "value": 40
  }
]
```

Finally, the pièce de résistance, the `metrics/plot` command to render the sampled metrics:
image

And there's also `metrics/html` which generates a static HTML page with the samples:
```
$ kubectl exec  -it -n kube-system ds/cilium -- cilium-dbg shell -- metrics/html > /tmp/out.html
```

2024-12-02-174915_hyprshot

You can find a demo from https://joamaki.github.io/demos/metrics.html

This html file is included in the sysdump in `cmd/cilium-dbg-shell----metrics-html.html`

### CODE REVIEW

**Review by learnitall — APPROVED** [review]
Vendor changes LGTM!

**Review by michi-covalent — APPROVED** [review]
cli changes looks good ✅ 📎

**Review by mhofstetter — APPROVED** [review]
`cilium/cli` :heavy_check_mark:

File: pkg/metrics/cmd.go [hunk 1]
```diff
@@ -47,13 +53,18 @@ func newMetricsCommand(r *Registry) hive.ScriptCmdOut {
...
 			},
 			func(s *script.State, args ...string) (script.WaitFunc, error) {
 				flags := flag.NewFlagSet("metrics", flag.ContinueOnError)
 				file := flags.String("o", "", "Output file")
 				format := flags.String("format", "table", "Output format, one of: table, json or yaml")
+				samples := flags.Bool("s", false, "Show sampled metrics")
 				if err := flags.Parse(args); err != nil {
+					if errors.Is(err, flag.ErrHelp) {
+
+					}
```
Comment (dylandreimerink) [inline pkg/metrics/cmd.go]: This seems to be missing something.

**Review by dylandreimerink — APPROVED** [review]
I did not see anything offending in the code, but this is at the point where I can't vouch for the correctness of the logic itself, its just to much at a time.

Found one small thing.

File: pkg/metrics/cmd.go (hunk 1 above)
Comment (dylandreimerink) [reply pkg/metrics/cmd.go]: Oh, nevermind, you fix it before I could finish the review.

**Review by ghost — APPROVED** [review]
nice

**Pull request merged**